### PR TITLE
Feature/add possibility to rent a bike

### PR DESCRIPTION
**Pull request by BartoliniAndBorderCollies**
#### Issue
https://trello.com/c/194XfL4m

### Description
This branch enables to `update` bike and its relations to user and charging station. It is possible to rent and return a bike and change all variables depending on a `role` (admin-user) in a application.

### Motivation
These functionalities are crucial to have a bike REST API.

### Changes introduced by this pull request
- add relations and annotations between entities (`@OneToOne`, `@OneToMany`, `@ManyToOne`)
- add `cascades` and `fetch.types`
- add annotation used in `Jackson` to process JSON data (`@JsonIdentityInfo`)
- add `Data Transer Object` for user
- add custom exception
- make separate endpoints and update methods depending on a role in app
- add validations so that even if client's JSON has some data it will not be executed on database if its requirement is not fulfilled
- make update method `@Transactional`

### CODE REVIEW

File: src/main/java/com/klodnicki/bike/rest/API/Bike/service/BikeService.java [hunk 1]
```diff
@@ -28,18 +39,86 @@ public Bike findBikeById(Long id) throws NotFoundInDatabaseException {
...
+                    user.setBike(null); // second side
+                    userService.addUser(user);  // Save the User to persist changes
+                }
+            }
+        }
         return bikeRepository.save(bike);
     }
+
+    public BikeForNormalUserDTO updateBikeUser(Long id, Bike bikeToUpdate) throws NotFoundInDatabaseException {
+        Bike bike = bikeRepository.findById(id).orElseThrow(NotFoundInDatabaseException::new);
```
Comment (JasinskiSz) [inline src/main/java/com/klodnicki/bike/rest/API/Bike/service/BikeService.java]: `Bike bike = bikeRepository.findById(id).orElseThrow(NotFoundInDatabaseException::new);`
Code repetition. There is already a method which does the same.

File: src/main/java/com/klodnicki/bike/rest/API/Bike/rest/controller/BikeController.java [hunk 2]
```diff
@@ -36,12 +41,38 @@ public void deleteBikeById(@PathVariable("id") Long id) throws NotFoundInDatabas
...
+        throw new UnauthorizedException();
     }
 
+    @PutMapping("/user/{id}")
+    public BikeForNormalUserDTO userUpdateBike(@PathVariable Long id, @RequestBody @Valid Bike bikeToUpdate,
+       @AuthenticationPrincipal UserDetails user) throws UnauthorizedException, NotFoundInDatabaseException {
 
+//        if(user.getUsername().equals("user")) {
+//            return bikeService.updateBikeUser(id, bikeToUpdate);
+//        }
```
Comment (JasinskiSz) [inline src/main/java/com/klodnicki/bike/rest/API/Bike/rest/controller/BikeController.java:67]: Keeping commented code is a bad practice!

**Review by JasinskiSz — APPROVED** [review]
Overall it looks ok

File: src/main/java/com/klodnicki/bike/rest/API/Bike/service/BikeService.java (hunk 1 above)
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/klodnicki/bike/rest/API/Bike/service/BikeService.java]: OK! I replace it! thanks for that!

File: src/main/java/com/klodnicki/bike/rest/API/Bike/rest/controller/BikeController.java (hunk 2 above)
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/klodnicki/bike/rest/API/Bike/rest/controller/BikeController.java:67]: I know its only for a practice workshop

**Pull request merged**